### ISSUE TITLE
Autopause being interrupted on an hourly basis

### ISSUE THREAD

**Issue by mhbates** [itzg/docker-minecraft-server#861]
Has anyone else experienced this? It started happening somewhat recently, probably in the last few months. Something is waking the server up about once an hour. I know that it could be something just knocking, but the Rcon line and the occasional bit about a client connecting (and no other info besides that, so likely not an actual client/player) has me thinking something else is going on. Any help is appreciated!

I think it's safe to ignore the HikariPool lines, that looks like it's just a thread watcher that one of my plugins uses.

EDIT: Looking at it again, it seems like the Rcon lines are just the final items before the server pauses again. So I guess my question is, how can I figure out what's knocking the server, and more importantly, why does it occasionally show a client as connecting, without any further information?

```
[11:12:06 WARN]: Can't keep up! Is the server overloaded? Running 3409884ms or 68197 ticks behind
[11:12:06 WARN]: [com.wfector.lib.hikari.pool.HikariPool] HikariPool-1 - Thread starvation or clock leap detected (housekeeper delta=56m51s399ms962µs170ns).
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - stopping
[11:14:08 INFO]: Thread RCON Client /127.0.0.1 started
[11:14:08 INFO]: [Rcon: Saved the game]
[11:14:08 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2021-05-11T11:14:08-0700] [Autopause] Pausing Java process
[12:12:21 WARN]: Can't keep up! Is the server overloaded? Running 3492368ms or 69847 ticks behind
[12:12:21 WARN]: [com.wfector.lib.hikari.pool.HikariPool] HikariPool-1 - Thread starvation or clock leap detected (housekeeper delta=58m14s84ms61µs724ns).
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - stopping
[12:14:22 INFO]: Thread RCON Client /127.0.0.1 started
[12:14:22 INFO]: [Rcon: Saved the game]
[12:14:22 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2021-05-11T12:14:22-0700] [Autopause] Pausing Java process
[13:08:41 WARN]: Can't keep up! Is the server overloaded? Running 3258930ms or 65178 ticks behind
[13:08:41 WARN]: [com.wfector.lib.hikari.pool.HikariPool] HikariPool-1 - Thread starvation or clock leap detected (housekeeper delta=54m20s510ms926µs815ns).
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] Client connected - waiting for disconnect
[Autopause loop] All clients disconnected - stopping in 300 seconds
[Autopause loop] No client reconnected - stopping
[13:14:03 INFO]: Thread RCON Client /127.0.0.1 started
[13:14:03 INFO]: [Rcon: Saved the game]
[13:14:03 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2021-05-11T13:14:03-0700] [Autopause] Pausing Java process
```

**Comment by itzg** [comment]
Do you happen to have an mc-backup container running alongside this?

**Comment by mhbates** [comment]
> 
> 
> Do you happen to have an mc-backup container running alongside this?

No, I don't. By the way, the server is whitelist only, and whenever a player does join, it shows that in the log, with their name and so on. The "Client connected" thing is puzzling.

**Comment by kocane** [comment]
@mhbates did you solve this? I seem to have the same issue.

**Comment by mhbates** [comment]
I did not unfortunately. I just ended up disabling autopause since the server was basically kept awake anyway.

**Comment by kocane** [comment]
@mhbates  Hm, alright. Shame, I don't mind lowering that idle CPU usage.

@itzg Do you have any idea how this can be troubleshooted? My log looks similar.
```
[2022-08-26T04:15:00+02:00] [Autopause] Pausing Java process
[04:25:24 WARN]: Can't keep up! Is the server overloaded? Running 624169ms or 12483 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
@mhbates  Hm, alright. Shame, I don't mind lowering that idle CPU usage.
@itzg Do you have any idea how this can be troubleshooted? My log looks like this:
[2022-08-26T04:15:00+02:00] [Autopause] Pausing Java process
[04:25:24 WARN]: Can't keep up! Is the server overloaded? Running 624169ms or 12483 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
[04:27:28 INFO]: Thread RCON Client /127.0.0.1 started
[04:27:28 INFO]: [Rcon: Saved the game]
[04:27:28 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T04:27:28+02:00] [Autopause] Pausing Java process
[05:19:03 WARN]: Can't keep up! Is the server overloaded? Running 3094787ms or 61895 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
[05:21:07 INFO]: Thread RCON Client /127.0.0.1 started
[05:21:07 INFO]: [Rcon: Saved the game]
[05:21:07 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T05:21:07+02:00] [Autopause] Pausing Java process
[06:12:47 WARN]: Can't keep up! Is the server overloaded? Running 3099477ms or 61989 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
[06:14:52 INFO]: Thread RCON Client /127.0.0.1 started
[06:14:52 INFO]: [Rcon: Saved the game]
[06:14:52 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T06:14:52+02:00] [Autopause] Pausing Java process
[07:06:15 WARN]: Can't keep up! Is the server overloaded? Running 3082985ms or 61659 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
[07:08:20 INFO]: Thread RCON Client /127.0.0.1 started
[07:08:20 INFO]: [Rcon: Saved the game]
[07:08:20 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T07:08:20+02:00] [Autopause] Pausing Java process
[07:59:43 WARN]: Can't keep up! Is the server overloaded? Running 3083291ms or 61665 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
[08:01:43 INFO]: Thread RCON Client /127.0.0.1 started
[08:01:43 INFO]: [Rcon: Saved the game]
[08:01:43 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T08:01:43+02:00] [Autopause] Pausing Java process
[08:12:45 WARN]: Can't keep up! Is the server overloaded? Running 661919ms or 13238 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[04:27:28 INFO]: Thread RCON Client /127.0.0.1 started
[04:27:28 INFO]: [Rcon: Saved the game]
[04:27:28 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T04:27:28+02:00] [Autopause] Pausing Java process
[05:19:03 WARN]: Can't keep up! Is the server overloaded? Running 3094787ms or 61895 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
[05:21:07 INFO]: Thread RCON Client /127.0.0.1 started
[05:21:07 INFO]: [Rcon: Saved the game]
[05:21:07 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T05:21:07+02:00] [Autopause] Pausing Java process
[06:12:47 WARN]: Can't keep up! Is the server overloaded? Running 3099477ms or 61989 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
[06:14:52 INFO]: Thread RCON Client /127.0.0.1 started
[06:14:52 INFO]: [Rcon: Saved the game]
[06:14:52 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T06:14:52+02:00] [Autopause] Pausing Java process
[07:06:15 WARN]: Can't keep up! Is the server overloaded? Running 3082985ms or 61659 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
[07:08:20 INFO]: Thread RCON Client /127.0.0.1 started
[07:08:20 INFO]: [Rcon: Saved the game]
[07:08:20 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T07:08:20+02:00] [Autopause] Pausing Java process
[07:59:43 WARN]: Can't keep up! Is the server overloaded? Running 3083291ms or 61665 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
[Autopause loop] No client connected since startup / knocked - pausing
[08:01:43 INFO]: Thread RCON Client /127.0.0.1 started
[08:01:43 INFO]: [Rcon: Saved the game]
[08:01:43 INFO]: Thread RCON Client /127.0.0.1 shutting down
[2022-08-26T08:01:43+02:00] [Autopause] Pausing Java process
[08:12:45 WARN]: Can't keep up! Is the server overloaded? Running 661919ms or 13238 ticks behind
[Autopause loop] Server was knocked - waiting for clients or timeout
```
I don't have any tools running and there's only one other guy who knows about the server.

**Comment by itzg** [comment]
I just noticed the address of the knocker can be passed to the resume script. So, I'll adjust the code to log that and that will help.

**Comment by kocane** [comment]
> I just noticed the address of the knocker can be passed to the resume script. So, I'll adjust the code to log that and that will help. 

Awesome, thanks. Is there a way for me to enable it already now?

**Comment by itzg** [comment]
Oh no, this issue fell of my radar. Thanks for the ping @kocane   and I'll post here when the debug info is available.

**Comment by itzg** [comment]
@kocane the latest image will now show the IP address of the knocker. If the origin was external, the address may not be useful due to NAT'ing within the docker network bridge.

**Comment by kocane** [comment]
> @kocane the latest image will now show the IP address of the knocker. If the origin was external, the address may not be useful due to NAT'ing within the docker network bridge.

Thanks! Works exactly like it should and I found the culprit.

**Comment by kocane** [comment]
@itzg I'm still seeing some random wakeups (but far fewer now) once in a while. Looking at the IPs it's typically port scanners. I don't suppose there's much to do about that aside from IP whitelisting in the firewall? Could the autopause function be made so it takes 2 knocks on the server, within a small time frame, before it reacts? Maybe it's just overcomplicating it.

**Comment by itzg** [comment]
> Could the autopause function be made so it takes 2 knocks on the server, within a small time frame, before it reacts? Maybe it's just overcomplicating it.

Yes could be done, but the usability of that seems annoying.

**Comment by samit43** [comment]
Hi, I'm having the exact same thing, how do I see the IP addresses?

**Comment by samit43** [comment]
Oh, I restarted the container, and it works now. Turns out it's someone from Paris (https://whatismyipaddress.com/ip/163.172.139.143) and has a history of trying to join/ping Minecraft server. I'm guessing this can't be fixed, but is there some sort of IP blacklist that the container can use that lies before the knock detector thing?

**Comment by RomanSonnik** [comment]
I dont have ip information in docker-compose container 
[02:17:49] [Server thread/WARN]: Can't keep up! Is the server overloaded? Running 3240ms or 64 ticks behind
mc-1  | [02:17:49] [RCON Client /0:0:0:0:0:0:0:1 #43/INFO]: Thread RCON Client /0:0:0:0:0:0:0:1 shutting down
mc-1  | [2024-07-15T02:17:49+00:00] [Autopause] Pausing Java process
mc-1  | [02:20:13] [Server thread/WARN]: Can't keep up! Is the server overloaded? Running 129193ms or 2583 ticks behind
mc-1  | [Autopause loop] Server was knocked - waiting for clients or timeout
mc-1  | [Autopause loop] No client connected since startup / knocked - pausing
mc-1  | [02:22:16] [RCON Listener #1/INFO]: Thread RCON Client /0:0:0:0:0:0:0:1 started

**Comment by kawazoe** [comment]
@itzg Hello. I know this issue has been closed for a while, but I'm getting the same problem and I don't get any info in the logs. Was this feature removed?

**Comment by itzg** [comment]
Yes, knockd was never correctly providing the remote IP address, so I removed it from logging since it was misleading.

**Comment by petitbonney** [comment]
I think I found a way to avoid this pause/unpause behavior.
As this is caused by rcon auto save:
```
[07:08:20 INFO]: Thread RCON Client /127.0.0.1 started
[07:08:20 INFO]: [Rcon: Saved the game]
```
We can disable it when there is no players:
```yaml
environment:
  # Auto-Pause
  ENABLE_AUTOPAUSE: true
  RCON_CMDS_FIRST_CONNECT: |-
    save-on
    save-all flush
  RCON_CMDS_LAST_DISCONNECT: |-
    save-all flush
    save-off
```

What do you think?

**Comment by itzg** [comment]
That's seems like a reasonable solution @petitbonney ; however, the line

```
[07:08:20 INFO]: [Rcon: Saved the game]
```

makes me wondering if you're using https://github.com/itzg/docker-mc-backup and it's doing the save operations via rcon to prepare for backup.

rcon itself is just a protocol-connection handler. It won't be initiating any operations by itself.

**Comment by petitbonney** [comment]
Ah yes, I am using the backup image. Here is my whole compose file: https://github.com/petitbonney/kanaloa-mc/blob/master/docker-compose.yml

You are right, even if I suspended the saving, I still have the same logs. I just read the mc-backup readme, what if I set ENABLE_SAVE_ALL to false in addition of RCON_CMDS_FIRST_CONNECT/RCON_CMDS_LAST_DISCONNECT config? I will try and keep you updated.

**Comment by kawazoe** [comment]
Small update. I have finally figured out what was happening in my case. Turns out intrusion detection was turned off on my security gateway; and I get around a dozen different people trying to port-scan my network every day. While most of those are looking for WordPress flaws, some included explicit checks for Minecraft ports. They were enough to knock the server multiple times a day.

If your server is publicly accessible, that might be something you want to take into consideration.
